src/CostFunction3D.cpp

```diff
@@ -800,33 +806,31 @@ void CostFunction3D::calcInnovation()
 
 void CostFunction3D::calcHTranspose(const real* yhat, real* Astate)
 {
-  integer j,n,m,k,ms,me;
+  uint64_t n,m;         // uint64_t
+  uint64_t j,begin,end; // uint32_t
   real tmp,val;
-	#pragma acc data present(yhat,Astate,mPtr,mVal,I2H,H)
-	{
-  	GPTLstart("CostFunction3D::calcHTranspose");
-  	#pragma omp parallel for private(n,k,ms,me,tmp,m,j,val)
-  	#pragma acc parallel loop gang vector vector_length(32) private(n,k,ms,me,tmp,m,j,val)
-  	for(n=0;n<nState;n++){
-    	ms = mPtr[n];
-    	me = mPtr[n+1];
-    	tmp = 0;
-    	if(me>ms){
-      	for (k=ms;k<me;k++){
-          m=mVal[k];
-          j=I2H[k];
-          //val = yhat[m] * obsVector[m*(7+varDim*derivDim)+1];
-          val = yhat[m] * obsData[m];
-          tmp += H[j] * val;
-       }
-    }
-    Astate[n]=tmp;
-  }
-  GPTLstop("CostFunction3D::calcHTranspose");
-	}
-}
 
+        #pragma acc data present(yhat,Astate)
+        {
+        GPTLstart("CostFunction3D::calcHTranspose");
+        // Multiply the state by the observation matrix
+        #pragma omp parallel for private(n,m,j,tmp,begin,end) //[9]
+        #pragma acc parallel loop vector gang vector_length(32) private(n,m,j,tmp,begin,end) reduction(+:tmp) //[9]
```

Comment:
If you want to apply `gang vector` here, then you do not need to add the OpenACC pragma at line 822 because that loop will run serially. If you want to run line 822 in parallel on the GPU, you may need to remove the `vector` clause here and add the `vector` clause at line 822 instead.

Reply:
Good point thanks for the catch.